dht/range_streamer.cc

```diff
@@ -245,6 +245,7 @@ future<> range_streamer::add_ranges(const sstring& keyspace_name, locator::vnode
 future<> range_streamer::stream_async() {
     _nr_ranges_remaining = nr_ranges_to_stream();
     _nr_total_ranges = _nr_ranges_remaining;
+    _token_metadata_ptr = nullptr;
```

Comment:
What prevents an an outdated and abandoned stream from bypassing fencing now?

Reply:
The idea is that the coordinator does not abandon streaming just like that. This execution belongs to migration sequence, and the coordinator waits for it before it moves on. 

The reason not to use erm tracking for this is that it is global and this would prevent migration concurrency by blocking barriers for other tablets.  

We could use it if we had fencing and barriers with tablet-granularity. But then split/merge becomes problematic. 

The waiting for RPCs is implemented like this.

We track them on replica-side here:
```c++
    std::unordered_map<locator::global_tablet_id, std::optional<shared_future<>>> _tablet_streaming;
```

When coordinator retries, replica-side RPC joins with the previous operation:

Streaming registers itself here:
```c++
    if (_tablet_streaming[tablet]) {
        slogger.debug("Streaming retry joining with existing session for tablet {}", tablet);
        co_await _tablet_streaming[tablet]->get_future();
        co_return;
    }

    auto async_gate_holder = _async_gate.hold();
    promise<> p;
    _tablet_streaming[tablet] = seastar::shared_future<>(p.get_future());
    auto erase_tablet_streaming = seastar::defer([&] {
        _tablet_streaming.erase(tablet);
    });
```

We're missing fencing there, which I'm fixing right now.

But the idea is that a successful execution will guarantee that there aren't and won't be any instances of this operation running which were triggered by the coordinator(s) earlier.

There are no aborts implemented yet, but they could be implemented by a special kind of retry which interrupts and joins with the existing fiber (in addition to fencing).

Reply:
> The idea is that the coordinator does not abandon streaming just like that. This execution belongs to migration sequence, and the coordinator waits for it before it moves on.

A coordinator may have initiated streaming, then died.

> 
> The reason not to use erm tracking for this is that it is global and this would prevent migration concurrency by blocking barriers for other tablets.

We certainly can't use the "hold erm until end of request" for very long running requests. But I propose "hold erm, listen for new erms, update to new erm if compatible in the token range and locality we're interested in, otherwise [keep holding to our erm|cancel the operation] (erase whatever is inapplicable).

So migration can check that the erms are compatible for this tablet replica (ignoring everything else). #1865  queries can just drop themselves, or upgrade if we take the time to do it (for single-partition queries it's easy).

> 
> We could use it if we had fencing and barriers with tablet-granularity. But then split/merge becomes problematic.
> 
> The waiting for RPCs is implemented like this.
> 
> We track them on replica-side here:
> 
> ```c++
>     std::unordered_map<locator::global_tablet_id, std::optional<shared_future<>>> _tablet_streaming;
> ```
> 
> When coordinator retries, replica-side RPC joins with the previous operation:

I guess this retry also works across leadership changes. 


> 
> Streaming registers itself here:
> 
> ```c++
>     if (_tablet_streaming[tablet]) {
>         slogger.debug("Streaming retry joining with existing session for tablet {}", tablet);
>         co_await _tablet_streaming[tablet]->get_future();
>         co_return;
>     }
> 
>     auto async_gate_holder = _async_gate.hold();
>     promise<> p;
>     _tablet_streaming[tablet] = seastar::shared_future<>(p.get_future());
>     auto erase_tablet_streaming = seastar::defer([&] {
>         _tablet_streaming.erase(tablet);
>     });
> ```
> 
> We're missing fencing there, which I'm fixing right now.
> 
> But the idea is that a successful execution will guarantee that there aren't and won't be any instances of this operation running which were triggered by the coordinator(s) earlier.
> 
> There are no aborts implemented yet, but they could be implemented by a special kind of retry which interrupts and joins with the existing fiber (in addition to fencing).



I concede it works. I like holding on to erm because it gives us a single mechanism for mechanically signalling a lock on topology. With dropping erm we need the cooperation of higher level entities.

Maybe I'm placing too much emphasis on this, but that's why I like it.

Reply:
> A coordinator may have initiated streaming, then died.

In that case the new coordinator takes over responsibility for either moving it to completion or canceling.

But I see your point. And it may indeed be simpler than inventing a new set of barriers and fencing. 

I implemented your suggestion in #15197, as https://github.com/scylladb/scylladb/pull/15197/commits/6a62aca3a919d8b35165a058fa0487f0e3ff9ad3 and https://github.com/scylladb/scylladb/pull/15197/commits/5cf035878d955255dc8a4d40436353a5c2351b86